src/runtime/hexagon_remote/remote.cpp

```diff
@@ -44,15 +95,239 @@ int halide_do_task(void *user_context, halide_task_t f, int idx,
                    uint8_t *closure) {
     return f(user_context, idx, closure);
 }
+void lock_work_queue() {
+    qurt_mutex_lock(&(work_queue.work_mutex));
+}
+void unlock_work_queue() {
+    qurt_mutex_unlock(&(work_queue.work_mutex));
+}
+void goto_work(work *owned_job) {
+    work myjob;
+    // You can work only if you get a lock on the work queue.
+    unsigned int tid = qurt_thread_get_id();
+    FARF(LOW, "HVX_TP: %d: goto_work: Trying to get a lock on the work queue\n", tid);
+
+    // ***********************
+    // *** Lock work queue ***
+    // ***********************
+    lock_work_queue();
+    FARF(LOW, "HVX_TP: %d: goto_work: got a lock on the work queue\n", tid);
+
+    // If I'm a job owner, then I was the thread that called
+    // do_par_for, and I should only stay in this function while there
+    // are active workers on the job (i.e. running()) or if the job
+    // hasn't been claimed entirely (!claimed()). If I'm a lowly worker
+    // thread, I should stay in this function as long as the work queue is running.
+    while (owned_job != NULL ? (owned_job->running() || !owned_job->claimed())
+           : work_queue.running()) {
+
+        FARF(LOW, "HVX_TP: %d: goto_work: In the main goto_work loop \n", tid);
+        int num_jobs = work_queue.num_jobs, job_idx;
+        work *jobs = work_queue.jobs, *orig_job;
+        bool job_found = false;
+        if (!owned_job) {
+            // If this threads doesn't own a job, it looks for one
+            // and tries to do it. If it cannot find a job, it goes
+            // to sleep.
+            if (num_jobs > 0) {
```

Comment:
I think it would help to make this a separate function, like find_empty_job is (find_nonempty_job? ;) )